src/interpreter/interpreter-non-det.ts

```diff
@@ -0,0 +1,706 @@
+  (context.runtime.environments[0] = environment)
+const popEnvironment = (context: Context) => context.runtime.environments.shift()
+const pushEnvironment = (context: Context, environment: Environment) =>
+  context.runtime.environments.unshift(environment)
+
+const getVariable = (context: Context, name: string) => {
+  let environment: Environment | null = context.runtime.environments[0]
+  while (environment) {
+    if (environment.head.hasOwnProperty(name)) {
+      if (environment.head[name] === HOISTED_BUT_NOT_YET_ASSIGNED) {
+        return handleRuntimeError(
+          context,
+          new errors.UnassignedVariable(name, context.runtime.nodes[0])
+        )
+      } else {
+        return environment.head[name]
+      }
+    } else {
+      environment = environment.tail
+    }
+  }
+  return handleRuntimeError(context, new errors.UndefinedVariable(name, context.runtime.nodes[0]))
+}
+
+const setVariable = (context: Context, name: string, value: any) => {
+  let environment: Environment | null = context.runtime.environments[0]
+  while (environment) {
+    if (environment.head.hasOwnProperty(name)) {
+      if (environment.head[name] === HOISTED_BUT_NOT_YET_ASSIGNED) {
+        break
+      }
+      const descriptors = Object.getOwnPropertyDescriptors(environment.head)
+      if (descriptors[name].writable) {
+        environment.head[name] = value
+        return undefined
+      }
+      return handleRuntimeError(
+        context,
+        new errors.ConstAssignment(context.runtime.nodes[0]!, name)
+      )
+    } else {
+      environment = environment.tail
+    }
+  }
+  return handleRuntimeError(context, new errors.UndefinedVariable(name, context.runtime.nodes[0]))
+}
+
+const checkNumberOfArguments = (
+  context: Context,
+  callee: Closure,
+  args: Value[],
+  exp: es.CallExpression
+) => {
+  if (callee.node.params.length !== args.length) {
+    return handleRuntimeError(
+      context,
+      new errors.InvalidNumberOfArguments(exp, callee.node.params.length, args.length)
+    )
+  }
+  return undefined
+}
+
+function* getArgs(context: Context, call: es.CallExpression) {
+  const args = []
+  for (const arg of call.arguments) {
+    args.push(yield* evaluate(arg, context))
+  }
+  return args
+}
+
+function* getAmbArgs(context: Context, call: es.CallExpression) {
+  const originalContext = cloneDeep(context)
+  for (const arg of call.arguments) {
+    yield* evaluate(arg, context)
+    assignIn(context, cloneDeep(originalContext)) // reset context
+  }
+}
+
+function transformLogicalExpression(node: es.LogicalExpression): es.ConditionalExpression {
+  if (node.operator === '&&') {
+    return conditionalExpression(node.left, node.right, literal(false), node.loc!)
+  } else {
+    return conditionalExpression(node.left, literal(true), node.right, node.loc!)
+  }
+}
+
+function* evaluateRequire(context: Context, call: es.CallExpression) {
+  // TODO: Throw an error if require does not have 0 arguments
```

Comment:
Just adding a comment to mark this TODO: We need to validate that `require` has exactly 1 argument